System/FilePath/Glob/Directory.hs

```diff
@@ -213,21 +214,57 @@ matchTypedAndGo opts (AnyDir n p:ps) path absPath = do
              p' = Pattern (unPattern p ++ [AnyNonPathSeparator])
 
          case unconditionalMatch || matchWith (matchOptions opts) p' path of
-              True | isDir  -> do
-                 contents <- getRecursiveContents
-                                (absPath ++ replicate n pathSeparator)
+              True | isDir -> do
+                 contents <- getRecursiveContents absPath
                  return $
                     -- foo**/ should match foo/ and nothing below it
                     -- relies on head contents == absPath
                     if null ps
-                       then (DL.singleton $ DL.head contents, DL.tail contents)
-                       else partitionDL (any m . pathParts) contents
-
-              True | m path -> return (DL.singleton absPath, DL.empty)
-              _             -> didn'tMatch opts path absPath isDir
+                       then ( DL.singleton $
+                                DL.head contents
+                                ++ replicate n pathSeparator
+                            , DL.tail contents
+                            )
+                       else let (matches, nonMatches) =
+                                   partitionDL fst
+                                      (fmap (recursiveMatch n m) contents)
+                             in (fmap snd matches, fmap snd nonMatches)
+
+              True | m path ->
+                 return ( DL.singleton $
+                             takeDirectory absPath
+                             ++ replicate n pathSeparator
+                             ++ path
+                        , DL.empty
+                        )
+              _ ->
+                 didn'tMatch opts path absPath isDir
 
 matchTypedAndGo _ _ _ _ = error "Glob.matchTypedAndGo :: internal error"
 
+recursiveMatch :: Int -> (FilePath -> Bool) -> FilePath -> (Bool, FilePath)
```

Comment:
This function could use a comment describing what it does, since the type signature isn't very self-explanatory.